lib/undo_redo.js

```diff
@@ -0,0 +1,121 @@
+(function (global) {
+    'use strict';
+    const fabric = global.fabric || (global.fabric = {});
+
+    const _saveObjectTransform = fabric.util.saveObjectTransform;
+    fabric.util.saveObjectTransform = function (target) {
+        const v = _saveObjectTransform(target);
+        v['width'] = target.width * target.scaleX;
+        v['height'] = target.height * target.scaleY;
+        ['rx', 'ry', 'radius', 'fill', 'stroke'].forEach(prop => prop in target && target[prop] != null && (v[prop] = target[prop]));
+        return v;
+    };
+
+    const _initialize = fabric.Canvas.prototype.initialize;
+    fabric.Canvas.prototype.initialize = function () {
+        const t = _initialize.call(this);
+
+        this.clearUndoRedo();
+        this.on('object:added', e => {
+            if (this.undoredo.escape || !e.target || e.target.excludeFromExport) {
+                return;
+            }
+            this.undoredo.stack[++this.undoredo.index] = { kind: 'added', target: e.target };
+            this.undoredo.stack.length = this.undoredo.index + 1;
+        });
+        this.on('object:removed', e => {
+            if (this.undoredo.escape || !e.target || e.target.excludeFromExport) {
+                return;
+            }
+            this.undoredo.stack[++this.undoredo.index] = { kind: 'removed', target: e.target };
+            this.undoredo.stack.length = this.undoredo.index + 1;
+        });
+        this.on('object:modified', e => {
+            if (this.undoredo.escape || !e.target || e.target.excludeFromExport) {
+                return;
+            }
+            const destination = Object.create(null);
+            Object.keys(e.transform.original).forEach(k => destination[k] = e.target[k], this);
+            this.undoredo.stack[++this.undoredo.index] = { kind: 'modified', action: e.action, target: e.target, original: e.transform.original, destination };
+            this.undoredo.stack.length = this.undoredo.index + 1;
+        });
+        this.on('canvas:cleared', _ => this.clearUndoRedo());
+        return t;
+    }
+
+    const ___setupCanvas = fabric.canvas.prototype.__setupCanvas;
+    fabric.canvas.prototype.__setupCanvas = function (serialized, enlivenedObjects, renderOnAddRemove, callback) {
```

Comment:
```suggestion
    const ___setupCanvas = fabric.Canvas.prototype.__setupCanvas;
    fabric.Canvas.prototype.__setupCanvas = function (serialized, enlivenedObjects, renderOnAddRemove, callback) {
```

Nice lib idea! `fabric.canvas` should have uppercase C though